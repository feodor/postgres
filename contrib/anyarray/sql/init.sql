CREATE EXTENSION anyarray;

--sanity check
SELECT 
	opc.opcname, 
	t.typname, 
	opc.opcdefault  
FROM 
	pg_opclass opc, 
	pg_am am, 
	pg_type t, 
	pg_type k 
WHERE 
	opc.opcmethod = am.oid AND 
	am.amname='gist' AND 
	opc.opcintype = t.oid AND 
	opc.opckeytype = k.oid AND 
	k.typname='ganyarray'
ORDER BY opc.opcname;

SELECT 
	opc.opcname, 
	t.typname, 
	opc.opcdefault  
FROM 
	pg_opclass opc, 
	pg_am am, 
	pg_type t
WHERE 
	opc.opcmethod = am.oid AND 
	am.amname='gin' AND 
	opc.opcintype = t.oid AND 
	opc.opcname ~ '_aa_ops$'
ORDER BY opc.opcname;

SELECT 
    trim( leading '_'  from t.typname ) || '[]' AS "Array Type",
    gin.opcname AS "GIN operator class",
    gist.opcname AS "GiST operator class"
FROM
    (
        SELECT *
        FROM
            pg_catalog.pg_opclass,
            pg_catalog.pg_am
        WHERE
            pg_opclass.opcmethod = pg_am.oid AND
            pg_am.amname = 'gin' AND
            pg_opclass.opcname ~ '_aa_ops$'
    ) AS gin
    FULL OUTER JOIN
        (
            SELECT *
            FROM
                pg_catalog.pg_opclass,
                pg_catalog.pg_am
            WHERE
                pg_opclass.opcmethod = pg_am.oid AND
                pg_am.amname = 'gist' AND
                pg_opclass.opcname ~ '_aa_ops$'
        ) AS gist
        ON (
            gist.opcname = gin.opcname AND 
            gist.opcintype = gin.opcintype 
        ),
    pg_catalog.pg_type t
WHERE
    t.oid = COALESCE(gist.opcintype, gin.opcintype) AND
    t.typarray = 0
ORDER BY
    "Array Type" ASC 
;

--testing function
CREATE OR REPLACE FUNCTION epoch2timestamp(int8)
RETURNS timestamp AS $$
    SELECT ('1970-01-01 00:00:00'::timestamp +   ( ($1*3600*24 + $1) ::text || ' seconds' )::interval)::timestamp;
	$$ LANGUAGE SQL RETURNS NULL ON NULL INPUT IMMUTABLE;

CREATE OR REPLACE FUNCTION to_tsp_array(_int8)
RETURNS _timestamp AS $$
	SELECT ARRAY( 
		SELECT 
			epoch2timestamp( $1[n] )
		FROM
			generate_series( 1, array_upper( $1, 1) - array_lower( $1, 1 ) + 1 ) AS n
	);
	$$ LANGUAGE SQL RETURNS NULL ON NULL INPUT IMMUTABLE;

CREATE OR REPLACE FUNCTION to_macp_array(_int8)
RETURNS _macaddr AS $$
	SELECT ARRAY( 
		SELECT 
			('01:01:01:01:01:' || to_hex($1[n] % 256))::macaddr	
		FROM
			generate_series( 1, array_upper( $1, 1) - array_lower( $1, 1 ) + 1 ) AS n
	);
	$$ LANGUAGE SQL RETURNS NULL ON NULL INPUT IMMUTABLE;

CREATE OR REPLACE FUNCTION to_inetp_array(_int8)
RETURNS _inet AS $$
	SELECT ARRAY( 
		SELECT 
			('192.168.1.' || ($1[n] % 256))::inet	
		FROM
			generate_series( 1, array_upper( $1, 1) - array_lower( $1, 1 ) + 1 ) AS n
	);
	$$ LANGUAGE SQL RETURNS NULL ON NULL INPUT IMMUTABLE;

CREATE OR REPLACE FUNCTION array_to_col(anyarray)
RETURNS SETOF anyelement AS
$$
    SELECT $1[n] FROM generate_series( 1, array_upper( $1, 1) - array_lower( $1, 1 ) + 1 ) AS n;
$$ LANGUAGE SQL RETURNS NULL ON NULL INPUT IMMUTABLE;

CREATE OR REPLACE FUNCTION max(int8, int8)
RETURNS int8 AS
$$
	SELECT CASE WHEN $1 > $2 THEN $1 ELSE $2 END;
$$ LANGUAGE SQL RETURNS NULL ON NULL INPUT IMMUTABLE;
