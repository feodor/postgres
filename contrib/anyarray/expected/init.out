CREATE EXTENSION anyarray;
--sanity check
SELECT 
	opc.opcname, 
	t.typname, 
	opc.opcdefault  
FROM 
	pg_opclass opc, 
	pg_am am, 
	pg_type t, 
	pg_type k 
WHERE 
	opc.opcmethod = am.oid AND 
	am.amname='gist' AND 
	opc.opcintype = t.oid AND 
	opc.opckeytype = k.oid AND 
	k.typname='ganyarray'
ORDER BY opc.opcname;
       opcname       |   typname    | opcdefault 
---------------------+--------------+------------
 _bytea_aa_ops       | _bytea       | f
 _char_aa_ops        | _char        | f
 _cidr_aa_ops        | _cidr        | f
 _date_aa_ops        | _date        | f
 _float4_aa_ops      | _float4      | f
 _float8_aa_ops      | _float8      | f
 _inet_aa_ops        | _inet        | f
 _int2_aa_ops        | _int2        | f
 _int4_aa_ops        | _int4        | f
 _int8_aa_ops        | _int8        | f
 _interval_aa_ops    | _interval    | f
 _macaddr_aa_ops     | _macaddr     | f
 _numeric_aa_ops     | _numeric     | f
 _oid_aa_ops         | _oid         | f
 _text_aa_ops        | _text        | f
 _time_aa_ops        | _time        | f
 _timestamp_aa_ops   | _timestamp   | f
 _timestamptz_aa_ops | _timestamptz | f
 _timetz_aa_ops      | _timetz      | f
 _varchar_aa_ops     | _varchar     | f
(20 rows)

SELECT 
	opc.opcname, 
	t.typname, 
	opc.opcdefault  
FROM 
	pg_opclass opc, 
	pg_am am, 
	pg_type t
WHERE 
	opc.opcmethod = am.oid AND 
	am.amname='gin' AND 
	opc.opcintype = t.oid AND 
	opc.opcname ~ '_aa_ops$'
ORDER BY opc.opcname;
       opcname       |   typname    | opcdefault 
---------------------+--------------+------------
 _bit_aa_ops         | _bit         | f
 _bytea_aa_ops       | _bytea       | f
 _char_aa_ops        | _char        | f
 _cidr_aa_ops        | _cidr        | f
 _date_aa_ops        | _date        | f
 _float4_aa_ops      | _float4      | f
 _float8_aa_ops      | _float8      | f
 _inet_aa_ops        | _inet        | f
 _int2_aa_ops        | _int2        | f
 _int4_aa_ops        | _int4        | f
 _int8_aa_ops        | _int8        | f
 _interval_aa_ops    | _interval    | f
 _macaddr_aa_ops     | _macaddr     | f
 _money_aa_ops       | _money       | f
 _numeric_aa_ops     | _numeric     | f
 _oid_aa_ops         | _oid         | f
 _text_aa_ops        | _text        | f
 _time_aa_ops        | _time        | f
 _timestamp_aa_ops   | _timestamp   | f
 _timestamptz_aa_ops | _timestamptz | f
 _timetz_aa_ops      | _timetz      | f
 _varbit_aa_ops      | _varbit      | f
 _varchar_aa_ops     | _varchar     | f
(23 rows)

SELECT
	am.amname,
	opc.opcname,
	o.oprname,
	count(*) over (partition by am.oid, opc.oid) as "N operators"
FROM
	pg_opclass opc,
	pg_am am,
	pg_amop amop,
	pg_operator o
WHERE
	opc.opcmethod = am.oid AND
	opc.opcname ~ '_aa_ops$' AND
	amop.amopfamily = opc.opcfamily AND
	amop.amopopr = o.oid
ORDER BY
	1, 2, 3;
 amname |       opcname       | oprname | N operators 
--------+---------------------+---------+-------------
 gin    | _bit_aa_ops         | %       |           5
 gin    | _bit_aa_ops         | &&      |           5
 gin    | _bit_aa_ops         | <@      |           5
 gin    | _bit_aa_ops         | =       |           5
 gin    | _bit_aa_ops         | @>      |           5
 gin    | _bytea_aa_ops       | %       |           5
 gin    | _bytea_aa_ops       | &&      |           5
 gin    | _bytea_aa_ops       | <@      |           5
 gin    | _bytea_aa_ops       | =       |           5
 gin    | _bytea_aa_ops       | @>      |           5
 gin    | _char_aa_ops        | %       |           5
 gin    | _char_aa_ops        | &&      |           5
 gin    | _char_aa_ops        | <@      |           5
 gin    | _char_aa_ops        | =       |           5
 gin    | _char_aa_ops        | @>      |           5
 gin    | _cidr_aa_ops        | %       |           5
 gin    | _cidr_aa_ops        | &&      |           5
 gin    | _cidr_aa_ops        | <@      |           5
 gin    | _cidr_aa_ops        | =       |           5
 gin    | _cidr_aa_ops        | @>      |           5
 gin    | _date_aa_ops        | %       |           5
 gin    | _date_aa_ops        | &&      |           5
 gin    | _date_aa_ops        | <@      |           5
 gin    | _date_aa_ops        | =       |           5
 gin    | _date_aa_ops        | @>      |           5
 gin    | _float4_aa_ops      | %       |           5
 gin    | _float4_aa_ops      | &&      |           5
 gin    | _float4_aa_ops      | <@      |           5
 gin    | _float4_aa_ops      | =       |           5
 gin    | _float4_aa_ops      | @>      |           5
 gin    | _float8_aa_ops      | %       |           5
 gin    | _float8_aa_ops      | &&      |           5
 gin    | _float8_aa_ops      | <@      |           5
 gin    | _float8_aa_ops      | =       |           5
 gin    | _float8_aa_ops      | @>      |           5
 gin    | _inet_aa_ops        | %       |           5
 gin    | _inet_aa_ops        | &&      |           5
 gin    | _inet_aa_ops        | <@      |           5
 gin    | _inet_aa_ops        | =       |           5
 gin    | _inet_aa_ops        | @>      |           5
 gin    | _int2_aa_ops        | %       |           5
 gin    | _int2_aa_ops        | &&      |           5
 gin    | _int2_aa_ops        | <@      |           5
 gin    | _int2_aa_ops        | =       |           5
 gin    | _int2_aa_ops        | @>      |           5
 gin    | _int4_aa_ops        | %       |           5
 gin    | _int4_aa_ops        | &&      |           5
 gin    | _int4_aa_ops        | <@      |           5
 gin    | _int4_aa_ops        | =       |           5
 gin    | _int4_aa_ops        | @>      |           5
 gin    | _int8_aa_ops        | %       |           5
 gin    | _int8_aa_ops        | &&      |           5
 gin    | _int8_aa_ops        | <@      |           5
 gin    | _int8_aa_ops        | =       |           5
 gin    | _int8_aa_ops        | @>      |           5
 gin    | _interval_aa_ops    | %       |           5
 gin    | _interval_aa_ops    | &&      |           5
 gin    | _interval_aa_ops    | <@      |           5
 gin    | _interval_aa_ops    | =       |           5
 gin    | _interval_aa_ops    | @>      |           5
 gin    | _macaddr_aa_ops     | %       |           5
 gin    | _macaddr_aa_ops     | &&      |           5
 gin    | _macaddr_aa_ops     | <@      |           5
 gin    | _macaddr_aa_ops     | =       |           5
 gin    | _macaddr_aa_ops     | @>      |           5
 gin    | _money_aa_ops       | %       |           5
 gin    | _money_aa_ops       | &&      |           5
 gin    | _money_aa_ops       | <@      |           5
 gin    | _money_aa_ops       | =       |           5
 gin    | _money_aa_ops       | @>      |           5
 gin    | _numeric_aa_ops     | %       |           5
 gin    | _numeric_aa_ops     | &&      |           5
 gin    | _numeric_aa_ops     | <@      |           5
 gin    | _numeric_aa_ops     | =       |           5
 gin    | _numeric_aa_ops     | @>      |           5
 gin    | _oid_aa_ops         | %       |           5
 gin    | _oid_aa_ops         | &&      |           5
 gin    | _oid_aa_ops         | <@      |           5
 gin    | _oid_aa_ops         | =       |           5
 gin    | _oid_aa_ops         | @>      |           5
 gin    | _text_aa_ops        | %       |           5
 gin    | _text_aa_ops        | &&      |           5
 gin    | _text_aa_ops        | <@      |           5
 gin    | _text_aa_ops        | =       |           5
 gin    | _text_aa_ops        | @>      |           5
 gin    | _time_aa_ops        | %       |           5
 gin    | _time_aa_ops        | &&      |           5
 gin    | _time_aa_ops        | <@      |           5
 gin    | _time_aa_ops        | =       |           5
 gin    | _time_aa_ops        | @>      |           5
 gin    | _timestamp_aa_ops   | %       |           5
 gin    | _timestamp_aa_ops   | &&      |           5
 gin    | _timestamp_aa_ops   | <@      |           5
 gin    | _timestamp_aa_ops   | =       |           5
 gin    | _timestamp_aa_ops   | @>      |           5
 gin    | _timestamptz_aa_ops | %       |           5
 gin    | _timestamptz_aa_ops | &&      |           5
 gin    | _timestamptz_aa_ops | <@      |           5
 gin    | _timestamptz_aa_ops | =       |           5
 gin    | _timestamptz_aa_ops | @>      |           5
 gin    | _timetz_aa_ops      | %       |           5
 gin    | _timetz_aa_ops      | &&      |           5
 gin    | _timetz_aa_ops      | <@      |           5
 gin    | _timetz_aa_ops      | =       |           5
 gin    | _timetz_aa_ops      | @>      |           5
 gin    | _varbit_aa_ops      | %       |           5
 gin    | _varbit_aa_ops      | &&      |           5
 gin    | _varbit_aa_ops      | <@      |           5
 gin    | _varbit_aa_ops      | =       |           5
 gin    | _varbit_aa_ops      | @>      |           5
 gin    | _varchar_aa_ops     | %       |           5
 gin    | _varchar_aa_ops     | &&      |           5
 gin    | _varchar_aa_ops     | <@      |           5
 gin    | _varchar_aa_ops     | =       |           5
 gin    | _varchar_aa_ops     | @>      |           5
 gist   | _bytea_aa_ops       | %       |           5
 gist   | _bytea_aa_ops       | &&      |           5
 gist   | _bytea_aa_ops       | <@      |           5
 gist   | _bytea_aa_ops       | =       |           5
 gist   | _bytea_aa_ops       | @>      |           5
 gist   | _char_aa_ops        | %       |           5
 gist   | _char_aa_ops        | &&      |           5
 gist   | _char_aa_ops        | <@      |           5
 gist   | _char_aa_ops        | =       |           5
 gist   | _char_aa_ops        | @>      |           5
 gist   | _cidr_aa_ops        | %       |           5
 gist   | _cidr_aa_ops        | &&      |           5
 gist   | _cidr_aa_ops        | <@      |           5
 gist   | _cidr_aa_ops        | =       |           5
 gist   | _cidr_aa_ops        | @>      |           5
 gist   | _date_aa_ops        | %       |           5
 gist   | _date_aa_ops        | &&      |           5
 gist   | _date_aa_ops        | <@      |           5
 gist   | _date_aa_ops        | =       |           5
 gist   | _date_aa_ops        | @>      |           5
 gist   | _float4_aa_ops      | %       |           5
 gist   | _float4_aa_ops      | &&      |           5
 gist   | _float4_aa_ops      | <@      |           5
 gist   | _float4_aa_ops      | =       |           5
 gist   | _float4_aa_ops      | @>      |           5
 gist   | _float8_aa_ops      | %       |           5
 gist   | _float8_aa_ops      | &&      |           5
 gist   | _float8_aa_ops      | <@      |           5
 gist   | _float8_aa_ops      | =       |           5
 gist   | _float8_aa_ops      | @>      |           5
 gist   | _inet_aa_ops        | %       |           5
 gist   | _inet_aa_ops        | &&      |           5
 gist   | _inet_aa_ops        | <@      |           5
 gist   | _inet_aa_ops        | =       |           5
 gist   | _inet_aa_ops        | @>      |           5
 gist   | _int2_aa_ops        | %       |           5
 gist   | _int2_aa_ops        | &&      |           5
 gist   | _int2_aa_ops        | <@      |           5
 gist   | _int2_aa_ops        | =       |           5
 gist   | _int2_aa_ops        | @>      |           5
 gist   | _int4_aa_ops        | %       |           5
 gist   | _int4_aa_ops        | &&      |           5
 gist   | _int4_aa_ops        | <@      |           5
 gist   | _int4_aa_ops        | =       |           5
 gist   | _int4_aa_ops        | @>      |           5
 gist   | _int8_aa_ops        | %       |           5
 gist   | _int8_aa_ops        | &&      |           5
 gist   | _int8_aa_ops        | <@      |           5
 gist   | _int8_aa_ops        | =       |           5
 gist   | _int8_aa_ops        | @>      |           5
 gist   | _interval_aa_ops    | %       |           5
 gist   | _interval_aa_ops    | &&      |           5
 gist   | _interval_aa_ops    | <@      |           5
 gist   | _interval_aa_ops    | =       |           5
 gist   | _interval_aa_ops    | @>      |           5
 gist   | _macaddr_aa_ops     | %       |           5
 gist   | _macaddr_aa_ops     | &&      |           5
 gist   | _macaddr_aa_ops     | <@      |           5
 gist   | _macaddr_aa_ops     | =       |           5
 gist   | _macaddr_aa_ops     | @>      |           5
 gist   | _numeric_aa_ops     | %       |           5
 gist   | _numeric_aa_ops     | &&      |           5
 gist   | _numeric_aa_ops     | <@      |           5
 gist   | _numeric_aa_ops     | =       |           5
 gist   | _numeric_aa_ops     | @>      |           5
 gist   | _oid_aa_ops         | %       |           5
 gist   | _oid_aa_ops         | &&      |           5
 gist   | _oid_aa_ops         | <@      |           5
 gist   | _oid_aa_ops         | =       |           5
 gist   | _oid_aa_ops         | @>      |           5
 gist   | _text_aa_ops        | %       |           5
 gist   | _text_aa_ops        | &&      |           5
 gist   | _text_aa_ops        | <@      |           5
 gist   | _text_aa_ops        | =       |           5
 gist   | _text_aa_ops        | @>      |           5
 gist   | _time_aa_ops        | %       |           5
 gist   | _time_aa_ops        | &&      |           5
 gist   | _time_aa_ops        | <@      |           5
 gist   | _time_aa_ops        | =       |           5
 gist   | _time_aa_ops        | @>      |           5
 gist   | _timestamp_aa_ops   | %       |           5
 gist   | _timestamp_aa_ops   | &&      |           5
 gist   | _timestamp_aa_ops   | <@      |           5
 gist   | _timestamp_aa_ops   | =       |           5
 gist   | _timestamp_aa_ops   | @>      |           5
 gist   | _timestamptz_aa_ops | %       |           5
 gist   | _timestamptz_aa_ops | &&      |           5
 gist   | _timestamptz_aa_ops | <@      |           5
 gist   | _timestamptz_aa_ops | =       |           5
 gist   | _timestamptz_aa_ops | @>      |           5
 gist   | _timetz_aa_ops      | %       |           5
 gist   | _timetz_aa_ops      | &&      |           5
 gist   | _timetz_aa_ops      | <@      |           5
 gist   | _timetz_aa_ops      | =       |           5
 gist   | _timetz_aa_ops      | @>      |           5
 gist   | _varchar_aa_ops     | %       |           5
 gist   | _varchar_aa_ops     | &&      |           5
 gist   | _varchar_aa_ops     | <@      |           5
 gist   | _varchar_aa_ops     | =       |           5
 gist   | _varchar_aa_ops     | @>      |           5
(215 rows)

SELECT 
    trim( leading '_'  from t.typname ) || '[]' AS "Array Type",
    gin.opcname AS "GIN operator class",
    gist.opcname AS "GiST operator class"
FROM
    (
        SELECT *
        FROM
            pg_catalog.pg_opclass,
            pg_catalog.pg_am
        WHERE
            pg_opclass.opcmethod = pg_am.oid AND
            pg_am.amname = 'gin' AND
            pg_opclass.opcname ~ '_aa_ops$'
    ) AS gin
    FULL OUTER JOIN
        (
            SELECT *
            FROM
                pg_catalog.pg_opclass,
                pg_catalog.pg_am
            WHERE
                pg_opclass.opcmethod = pg_am.oid AND
                pg_am.amname = 'gist' AND
                pg_opclass.opcname ~ '_aa_ops$'
        ) AS gist
        ON (
            gist.opcname = gin.opcname AND 
            gist.opcintype = gin.opcintype 
        ),
    pg_catalog.pg_type t
WHERE
    t.oid = COALESCE(gist.opcintype, gin.opcintype) AND
    t.typarray = 0
ORDER BY
    "Array Type" ASC 
;
  Array Type   | GIN operator class  | GiST operator class 
---------------+---------------------+---------------------
 bit[]         | _bit_aa_ops         | 
 bytea[]       | _bytea_aa_ops       | _bytea_aa_ops
 char[]        | _char_aa_ops        | _char_aa_ops
 cidr[]        | _cidr_aa_ops        | _cidr_aa_ops
 date[]        | _date_aa_ops        | _date_aa_ops
 float4[]      | _float4_aa_ops      | _float4_aa_ops
 float8[]      | _float8_aa_ops      | _float8_aa_ops
 inet[]        | _inet_aa_ops        | _inet_aa_ops
 int2[]        | _int2_aa_ops        | _int2_aa_ops
 int4[]        | _int4_aa_ops        | _int4_aa_ops
 int8[]        | _int8_aa_ops        | _int8_aa_ops
 interval[]    | _interval_aa_ops    | _interval_aa_ops
 macaddr[]     | _macaddr_aa_ops     | _macaddr_aa_ops
 money[]       | _money_aa_ops       | 
 numeric[]     | _numeric_aa_ops     | _numeric_aa_ops
 oid[]         | _oid_aa_ops         | _oid_aa_ops
 text[]        | _text_aa_ops        | _text_aa_ops
 time[]        | _time_aa_ops        | _time_aa_ops
 timestamp[]   | _timestamp_aa_ops   | _timestamp_aa_ops
 timestamptz[] | _timestamptz_aa_ops | _timestamptz_aa_ops
 timetz[]      | _timetz_aa_ops      | _timetz_aa_ops
 varbit[]      | _varbit_aa_ops      | 
 varchar[]     | _varchar_aa_ops     | _varchar_aa_ops
(23 rows)

--testing function
CREATE OR REPLACE FUNCTION epoch2timestamp(int8)
RETURNS timestamp AS $$
    SELECT ('1970-01-01 00:00:00'::timestamp +   ( ($1*3600*24 + $1) ::text || ' seconds' )::interval)::timestamp;
	$$ LANGUAGE SQL RETURNS NULL ON NULL INPUT IMMUTABLE;
CREATE OR REPLACE FUNCTION to_tsp_array(_int8)
RETURNS _timestamp AS $$
	SELECT ARRAY( 
		SELECT 
			epoch2timestamp( $1[n] )
		FROM
			generate_series( 1, array_upper( $1, 1) - array_lower( $1, 1 ) + 1 ) AS n
	);
	$$ LANGUAGE SQL RETURNS NULL ON NULL INPUT IMMUTABLE;
CREATE OR REPLACE FUNCTION to_macp_array(_int8)
RETURNS _macaddr AS $$
	SELECT ARRAY( 
		SELECT 
			('01:01:01:01:01:' || to_hex($1[n] % 256))::macaddr	
		FROM
			generate_series( 1, array_upper( $1, 1) - array_lower( $1, 1 ) + 1 ) AS n
	);
	$$ LANGUAGE SQL RETURNS NULL ON NULL INPUT IMMUTABLE;
CREATE OR REPLACE FUNCTION to_inetp_array(_int8)
RETURNS _inet AS $$
	SELECT ARRAY( 
		SELECT 
			('192.168.1.' || ($1[n] % 256))::inet	
		FROM
			generate_series( 1, array_upper( $1, 1) - array_lower( $1, 1 ) + 1 ) AS n
	);
	$$ LANGUAGE SQL RETURNS NULL ON NULL INPUT IMMUTABLE;
CREATE OR REPLACE FUNCTION array_to_col(anyarray)
RETURNS SETOF anyelement AS
$$
    SELECT $1[n] FROM generate_series( 1, array_upper( $1, 1) - array_lower( $1, 1 ) + 1 ) AS n;
$$ LANGUAGE SQL RETURNS NULL ON NULL INPUT IMMUTABLE;
CREATE OR REPLACE FUNCTION max(int8, int8)
RETURNS int8 AS
$$
	SELECT CASE WHEN $1 > $2 THEN $1 ELSE $2 END;
$$ LANGUAGE SQL RETURNS NULL ON NULL INPUT IMMUTABLE;
